### PR TITLE
Manage before middleware error

### PR DESCRIPTION
**Pull request by ilkosta**
Hi, 

I added 

* a message to help the user running the example
* an around middleware to trace on the console the error message catched by the server

### CODE REVIEW

File: examples/error.rs [hunk 1]
```diff
@@ -37,17 +37,63 @@ impl Handler for ErrorHandler {
...
+
+// to handle the Error message by using an AroundMiddleware
+struct LogEnabler;
+
+struct HandlerWithLog<H: Handler> {
+    handler: H
+}
+
+impl AroundMiddleware for LogEnabler {
+
```
Comment (reem) [inline examples/error.rs]: delete this newline

File: examples/error.rs [hunk 2]
```diff
@@ -37,17 +37,63 @@ impl Handler for ErrorHandler {
...
+    }
+}
+ 
+impl <H: Handler> Handler for HandlerWithLog<H> {
+
+    fn handle(&self, req: &mut Request) -> IronResult<Response> {
+    
+        let res = self.handler.handle(req);
+        
+        {
```
Comment (reem) [inline examples/error.rs]: why does this need to be in its own block?

File: examples/error.rs [hunk 3]
```diff
@@ -37,17 +37,63 @@ impl Handler for ErrorHandler {
...
+          match res {
+            Ok(_) => panic!("is not possible!"),
+            Err(IronError {error: ref what_went_wrong, ref response }) => println!("what went wrong: {:?} - response: {:?}", what_went_wrong, response)
```
Comment (reem) [inline examples/error.rs]: if you just do `IronError { ref error, ref response }` I think that will be clearer

File: examples/error.rs [hunk 4]
```diff
@@ -37,17 +37,63 @@ impl Handler for ErrorHandler {
...
+impl AroundMiddleware for LogEnabler {
+
+    fn around(self, handler: Box<Handler>) -> Box<Handler> {
+        Box::new(HandlerWithLog { handler: handler } ) as Box<Handler>
+    }
+}
+ 
+impl <H: Handler> Handler for HandlerWithLog<H> {
+
+    fn handle(&self, req: &mut Request) -> IronResult<Response> {
```
Comment (reem) [inline examples/error.rs]: I think the best way to write this method would be something like:

```rust
self.handler.handle(req).map_err(|e| println!(..); e)
```

File: examples/error.rs [hunk 5]
```diff
@@ -37,17 +37,63 @@ impl Handler for ErrorHandler {
...
+    
     let mut chain = Chain::new(ErrorHandler);
 
     // Link our error maker.
     chain.link_before(ErrorProducer);
-
-    Iron::new(chain).http("localhost:3000").unwrap();
+    
+    // enable the tracing of the error by using an AroundMiddleware
+    let logged_handler = get_log_enabled_handler(Box::new(chain));
```
Comment (reem) [inline examples/error.rs]: why not just `logged_handler = LogEnable.around(Box::new(chain))`

File: examples/error.rs [hunk 6]
```diff
@@ -37,17 +37,63 @@ impl Handler for ErrorHandler {
...
+    
+    println!("server running at http://localhost:3000");
+    println!("try it by opening the url in a browser or by: 'curl -v http://localhost:3000', 'curl -v http://localhost:3000/any_other_destination/same_bad_request'");
```
Comment (reem) [inline examples/error.rs]: better as documentation/comments then prints I think

**Comment by reem** [comment]
I left some specific comments, but at a higher level you actually do not need to use an AroundMiddleware here, an AfterMiddleware would do perfectly fine, since all you need to do is introspect the response.

**Comment by ilkosta** [comment]
Really thanks! 
This is better than I ever hoped.
As you have guessed I'm trying to learn rust using an Iron project as gym, and my pull is for a review :)

I missed completely `map_err` as well as the method invocation of a struct. I thought that the structs went always instantiated and binded for the `ref.method() syntax`.

Now I see that an AroundMiddleware was not necessary. Apart from this example I am interested in the use case of a rest webservice that uses json and wants to "describe" errors for not met requirements. In that case, with an AfterMiddleweare I cannot change the response content, only the response headers. True?

**Comment by reem** [comment]
@ilkosta actually you can change the response content, or even replace the response wholesale - nothing is written to the actual stream until the top-level Handler has returned (which necessitates all middleware having run).

File: examples/error.rs [hunk 7]
```diff
@@ -37,17 +36,32 @@ impl Handler for ErrorHandler {
 
 impl BeforeMiddleware for ErrorProducer {
     fn before(&self, _: &mut Request) -> IronResult<()> {
-        Err(IronError::new(StringError("Error".to_string()), status::BadRequest))
+        Err(IronError::new(StringError("You cannot access the resource!".to_string()), status::BadRequest))
+    }
+}
+
+// the error is simply printed on stdout and returned as is
+struct ErrorHandler;
```
Comment (reem) [inline examples/error.rs]: add a newline after this declaration

File: examples/error.rs [hunk 8]
```diff
@@ -37,17 +36,32 @@ impl Handler for ErrorHandler {
 
 impl BeforeMiddleware for ErrorProducer {
     fn before(&self, _: &mut Request) -> IronResult<()> {
-        Err(IronError::new(StringError("Error".to_string()), status::BadRequest))
+        Err(IronError::new(StringError("You cannot access the resource!".to_string()), status::BadRequest))
+    }
+}
+
+// the error is simply printed on stdout and returned as is
```
Comment (reem) [inline examples/error.rs]: Generally speaking can you make new comments full sentences with capital letters, punctuation, etc.? It makes the examples easier to read for future new users.

File: examples/error.rs [hunk 9]
```diff
@@ -37,17 +36,32 @@ impl Handler for ErrorHandler {
...
+    let mut chain = Chain::new(NeverCalledHandler);
 
     // Link our error maker.
     chain.link_before(ErrorProducer);
 
+    // enable the tracing of the error by catching it with an AfterMiddleware
+    chain.link_after(ErrorHandler);
+
+    println!("server running at http://localhost:3000");
+    println!("try it by opening the url in a browser or by: 'curl -v http://localhost:3000', 'curl -v http://localhost:3000/any_other_destination/same_bad_request'");
```
Comment (reem) [inline examples/error.rs]: I would rather these prints were replaced by comments.

File: examples/error.rs [hunk 10]
```diff
@@ -37,17 +36,32 @@ impl Handler for ErrorHandler {
...
-        Err(IronError::new(StringError("Error".to_string()), status::BadRequest))
+        Err(IronError::new(StringError("You cannot access the resource!".to_string()), status::BadRequest))
+    }
+}
+
+// the error is simply printed on stdout and returned as is
+struct ErrorHandler;
+impl AfterMiddleware for ErrorHandler {
+    fn catch(&self, _: &mut Request, err: IronError) -> IronResult<Response> {
+      println!("handled the error: {:?}",err.error);
```
Comment (reem) [inline examples/error.rs]: How about `println!("Request Error! Error was: {:?}", err.error)`?

**Comment by reem** [comment]
Closing due to inactivity but feel free to reopen with the requested changes.

**Pull request closed without merging**